cmds/core/ip/ip_linux.go

```diff
@@ -333,62 +333,70 @@ func routedel() error {
 	return nil
 }
 
-func route() error {
+func route(w io.Writer) error {
 	cursor++
 	if len(arg[cursor:]) == 0 {
-		return routeshow()
+		return routeshow(w)
 	}
 
 	whatIWant = []string{"show", "add", "del"}
 	switch one(arg[cursor], whatIWant) {
 	case "add":
-		return routeadd()
+		return routeadd(w)
 	case "del":
 		return routedel()
 	case "show":
-		return routeshow()
+		return routeshow(w)
 	}
 	return usage()
 }
 
-func main() {
+func run(out io.Writer) error {
 	// When this is embedded in busybox we need to reinit some things.
 	whatIWant = []string{"address", "route", "link", "neigh"}
 	cursor = 0
-	flag.Parse()
-	arg = flag.Args()
 
-	defer func() {
+	defer func() error {
 		switch err := recover().(type) {
 		case nil:
 		case error:
 			if strings.Contains(err.Error(), "index out of range") {
-				log.Fatalf("Args: %v, I got to arg %v, I wanted %v after that", arg, cursor, whatIWant)
+				return fmt.Errorf("Args: %v, I got to arg %v, I wanted %v after that", arg, cursor, whatIWant)
 			} else if strings.Contains(err.Error(), "slice bounds out of range") {
-				log.Fatalf("Args: %v, I got to arg %v, I wanted %v after that", arg, cursor, whatIWant)
+				return fmt.Errorf("Args: %v, I got to arg %v, I wanted %v after that", arg, cursor, whatIWant)
 			}
-			log.Fatalf("Bummer: %v", err)
+			return fmt.Errorf("Bummer: %v", err)
 		default:
-			log.Fatalf("unexpected panic value: %T(%v)", err, err)
+			return fmt.Errorf("unexpected panic value: %T(%v)", err, err)
 		}
+		return nil
 	}()
 
 	// The ip command doesn't actually follow the BNF it prints on error.
 	// There are lots of handy shortcuts that people will expect.
 	var err error
 	switch one(arg[cursor], whatIWant) {
 	case "address":
-		err = addrip()
+		err = addrip(out)
 	case "link":
-		err = link()
+		err = link(out)
 	case "route":
-		err = route()
+		err = route(out)
 	case "neigh":
-		err = neigh()
+		err = neigh(out)
 	default:
 		err = usage()
 	}
 	if err != nil {
-		log.Fatal(err)
+		return err
+	}
+	return nil
+}
+
+func main() {
+	flag.Parse()
+	arg = os.Args[1:]
+	if err := run(os.Stdout); err != nil {
+		log.Fatalf("ip: %v", err)
```

Comment:
The standard command does not print the ip: part. I wonder if, for automation, we should just print the error?

Reply:
Putting the command name first is a convention already established in some of the other commands.
But i don't feel strongly about this.